README.md
# color_picker_palette 0.1.3

A color picker palette with custom colors.

## Usage
Using `color_picker_palette` you can provide
user with list of custom colors to choose.

It will show a horizontal list of colors including
a button to add custom colors for user.

Long press will remove color from the list.

You can refer [color_picker_palette/example](https://github.com/vemarav/color_picker_palette/tree/master/example)

```dart
  colorPickerPalette = new ColorPickerPalette(
    size: 30.0,
    onChange: setColorState,
    colors: colors,
    buttonType: ButtonType.circular,
  );
  
  setColorState() {
    setState(() {
      color = colorPickerPalette.color();
    });
  }
```

## Contributing

We encourage you contribute to color_picker_palette flutter package.

## License

color_picker_palette is released under [MIT License](https://opensource.org/licenses/MIT)

## Social
[![Twitter Follow](https://img.shields.io/twitter/follow/vemarav.svg?style=social&label=Follow)](https://twitter.com/vemarav)



CHANGELOG.md
## 0.1.3

- Initial release


test/color_picker_palette_test.dart
import 'package:test/test.dart';

import 'package:color_picker_palette/color_picker_palette.dart';

void main() {
  test('adds one to input values', () {
    final calculator = new Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:color_picker_palette/color_picker_palette.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  ColorPickerPalette colorPickerPalette;
  Color color;
  List<Color> colors = [];

  @override
  void initState() {
    super.initState();
    colors.add(Theme.of(context).primaryColor);
    colors.add(Colors.black);
    colors.add(Colors.white);
    color = colors[0];
    colorPickerPalette = new ColorPickerPalette(
      size: 30.0,
      onChange: setColorState,
      colors: colors,
      buttonType: ButtonType.circular,
    );
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          backgroundColor: color,
          title: Text(
            'Color Picker Palette',
            style: TextStyle(color: Color(color.value - 10010110)),
          ),
        ),
        body: new Container(
          constraints: BoxConstraints.expand(),
          color: color,
          child: new Column(
            children: <Widget>[
              colorPickerPalette,
              Container(
                padding: EdgeInsets.all(16.0),
                child: Text(
                  colorPickerPalette.color().toString(),
                  style: TextStyle(
                    fontSize: 24.0,
                    fontWeight: FontWeight.w400,
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  setColorState() {
    setState(() {
      color = colorPickerPalette.color();
    });
  }
}


lib/circle_button.dart
import 'package:flutter/material.dart';

class CircleButton extends StatelessWidget {
  CircleButton({this.color, this.onTap, this.size, this.child})
      : assert(
  color != null,
  "Provide Color for CircleButton \n Example: " +
    "\n\n CircleButton(color: Colors.blue)");

  final Function onTap;
  final double size;
  final Widget child;
  final Color color;
  static const double DEFAULT_SIZE = 50.0;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        child: child,
        margin: EdgeInsets.all(4.0),
        width: size ?? DEFAULT_SIZE,
        height: size ?? DEFAULT_SIZE,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: color,
          border: Border.all(
            color: Colors.grey,
            width: 1.5,
          ),
        ),
      ),
    );
  }
}


lib/picker_widget.dart
import 'custom_button.dart';
import 'color_picker_palette.dart';
import 'package:flutter/material.dart';

class ColorPickerPaletteWidget extends StatefulWidget {
  ColorPickerPaletteWidget(
      {Key key, this.size, this.callback, this.colors, this.buttonType})
      : super(key: key);
  final double size;
  final List<Color> colors;
  final VoidCallback callback;
  final ButtonType buttonType;

  @override
  State<StatefulWidget> createState() => ColorPickerPaletteWidgetState(
        size: size,
        callback: callback,
        colors: colors,
        buttonType: buttonType,
      );
}

class ColorPickerPaletteWidgetState extends State<ColorPickerPaletteWidget> {
  ColorPickerPaletteWidgetState(
      {this.size, this.callback, this.colors, this.buttonType});

  final VoidCallback callback;
  final List<Color> colors;
  final double size;
  final ButtonType buttonType;
  static const double DEFAULT_SIZE = 50.0;
  Color color;
  _AddColorDialog _addColorDialog;
  GlobalKey<_AddColorDialogState> _addColorDialogKey = GlobalKey();

  @override
  void initState() {
    super.initState();
    _addColorDialog = _AddColorDialog(
      key: _addColorDialogKey,
      onChange: onDialogResult,
    );
  }

  @override
  Widget build(BuildContext context) {
    var containerSize = (size ?? DEFAULT_SIZE) + 16.0;
    return Container(
      height: containerSize,
      child: ListView(
        padding: EdgeInsets.fromLTRB(8.0, 0.0, 8.0, 0.0),
        scrollDirection: Axis.horizontal,
        children: circleButtons(),
      ),
    );
  }

  List<Widget> circleButtons() {
    List<Widget> buttonWidgets = [];
    colors.forEach((color) {
      buttonWidgets.add(
        CustomButton(
          buttonType: buttonType,
          color: color,
          onTap: () {
            setState(() {
              this.color = color;
            });
            invokeCallback();
          },
          size: size,
          onLongPress: () {
            if (this.color == color) {
              setState(() {
                this.color = colors[0];
                colors.remove(color);
              });
            } else {
              setState(() {
                colors.remove(color);
              });
            }
            invokeCallback();
          },
        ),
      );
    });
    buttonWidgets.add(
      CustomButton(
        buttonType: buttonType,
        size: size,
        child: Icon(
          Icons.add,
          color: Colors.grey.shade700,
        ),
        color: Colors.white,
        onTap: () {
          showColorDialog(context);
        },
      ),
    );
    return buttonWidgets;
  }

  Future<String> showColorDialog(BuildContext context) {
    return showDialog(
      context: context,
      builder: (BuildContext context) {
        return _addColorDialog;
      },
    );
  }

  void onDialogResult() {
    setState(() {
      color = _addColorDialogKey.currentState.color;
      colors.add(color);
    });
    invokeCallback();
  }

  void invokeCallback() {
    if (callback != null) {
      callback();
    } else {
      print("warning: onChange Callback Not implemented"
          "You can provide method #onChange as\n"
          "  onChangeColor() {"
          "    TODO// setState"
          "  }"
          "\n"
          "_picker = ColorPickerPalette(..., onChange: onChangeColor)");
    }
  }
}

class _AddColorDialog extends StatefulWidget {
  _AddColorDialog({Key key, this.onChange}) : super(key: key);
  final VoidCallback onChange;

  @override
  State<StatefulWidget> createState() =>
      _AddColorDialogState(callback: onChange);
}

class _AddColorDialogState extends State<_AddColorDialog> {
  _AddColorDialogState({this.callback});
  final VoidCallback callback;
  Color color;
  TextEditingController _textController;
  String _error;

  @override
  void initState() {
    super.initState();
    _textController = TextEditingController(text: '');
  }

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: alertTitle(),
      content: alertContent(),
      actions: <Widget>[
        okButton(),
        cancelButton(),
      ],
    );
  }

  Widget alertTitle() {
    return Text('Enter HEX color code');
  }

  Widget spanHash() {
    return Container(
      padding: EdgeInsets.all(8.0),
      decoration: BoxDecoration(
        color: Colors.grey.shade100,
        border: Border.all(
          color: Colors.grey.shade200,
          width: 2.0,
        ),
      ),
      child: Text(
        '#',
        style: TextStyle(
          color: Colors.black,
          fontSize: 20.0,
          fontWeight: FontWeight.w600,
        ),
      ),
    );
  }

  Widget spanInput() {
    return Expanded(
      child: Container(
        padding: EdgeInsets.all(8.0),
        decoration: BoxDecoration(
          color: Colors.grey.shade100,
          border: Border.all(
            color: Colors.grey.shade200,
            width: 2.0,
          ),
        ),
        child: EditableText(
          controller: _textController,
          cursorColor: Colors.black,
          cursorWidth: 2.0,
          focusNode: FocusNode(),
          style: TextStyle(
            color: Colors.black,
            fontSize: 20.0,
            fontWeight: FontWeight.w600,
          ),
        ),
      ),
    );
  }

  Widget alertContent() {
    return Container(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          Row(
            children: <Widget>[
              spanHash(),
              spanInput(),
            ],
          ),
          errorWidget(),
        ],
      ),
    );
  }

  Widget errorWidget() {
    if (_error != null) {
      return Container(
        padding: EdgeInsets.fromLTRB(0.0, 8.0, 0.0, 8.0),
        child: Text(
          _error ?? '',
          style: TextStyle(
            color: Colors.red.shade700,
          ),
        ),
      );
    } else {
      return Container();
    }
  }

  Widget okButton() {
    return FlatButton(
      child: Text(
        'OK',
        style: TextStyle(
          color: Colors.black,
        ),
      ),
      onPressed: onSaveColor,
    );
  }

  Widget cancelButton() {
    return FlatButton(
      child: Text(
        'CANCEL',
        style: TextStyle(
          color: Colors.black,
        ),
      ),
      onPressed: onCancelColor,
    );
  }

  void onSaveColor() {
    try {
      if (_textController.text.length != 6) throw FormatException;
      Color _color = new Color(int.parse("0xFF${_textController.text}"));
      setState(() {
        color = _color;
        _error = null;
      });
      callback();
      Navigator.pop(context);
    } catch (exception) {
      setState(() {
        _error = 'Input is invalid';
      });
    }
  }

  void onCancelColor() {
    Navigator.pop(context);
  }
}


lib/color_picker_palette.dart
library color_picker_palette;

import 'package:flutter/material.dart';
import 'package:color_picker_palette/picker_widget.dart';

enum ButtonType { circular, rounded, rectangle }

class ColorPickerPalette extends StatelessWidget {
  ColorPickerPalette({this.size, this.onChange, this.colors, this.buttonType})
      : assert(
            colors != null,
            "-------------------------------------------------"
            "\n\nError\n"
            "\tPlease provide colors property in Constructor\n"
            "\t\tExample\n"
            "\t\t1. picker = ColorPickerPalette("
            "\t\t\t\tcolors: [Colors.white, Colors.blue, Colors.red, Colors.green]\n"
            "\t\t\t);\n"
            "\t\t2. picker = ColorPickerPalette(colors: Colors.primaries)\n\n"
            "-------------------------------------------------"),
        assert(
            colors.isNotEmpty,
            "-------------------------------------------------"
            "\n\nError\n"
            "\tProvide at least one color in List\n"
            "\t\tpicker = ColorPickerPaletter(colors: [Colors.blue]);\n"
            "-------------------------------------------------");
  final GlobalKey<ColorPickerPaletteWidgetState> _widgetKey = GlobalKey();
  final double size;
  final List<Color> colors;
  final VoidCallback onChange;
  final ButtonType buttonType;

  @override
  Widget build(BuildContext context) {
    return ColorPickerPaletteWidget(
      key: _widgetKey,
      size: size,
      callback: onChange,
      colors: colors,
      buttonType: buttonType,
    );
  }

  Color color() {
    return _widgetKey.currentState?.color ?? colors[0];
  }
}


lib/custom_button.dart
import 'color_picker_palette.dart';
import 'package:flutter/material.dart';

class CustomButton extends StatelessWidget {
  CustomButton(
      {this.color,
      this.onTap,
      this.onLongPress,
      this.size,
      this.child,
      this.buttonType})
      : assert(
            color != null,
            "Provide Color for CircleButton \n Example: " +
                "\n\n CircleButton(color: Colors.blue)");

  final VoidCallback onTap;
  final VoidCallback onLongPress;
  final ButtonType buttonType;
  final double size;
  final Widget child;
  final Color color;
  static const String DEFAULT_TYPE = 'circular';
  static const double DEFAULT_SIZE = 50.0;

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap,
      child: Container(
        child: child,
        margin: EdgeInsets.all(4.0),
        width: size ?? DEFAULT_SIZE,
        height: (size ?? DEFAULT_SIZE),
        decoration: selected(),
      ),
      onLongPress: onLongPress,
    );
  }

  Decoration selected() {
    switch (buttonType) {
      case ButtonType.circular:
        return circular();
      case ButtonType.rounded:
        return rounded();
      case ButtonType.rectangle:
        return rectangular();
      default:
        return circular();
    }
  }

  Decoration rectangular() {
    return BoxDecoration(
      color: color,
      border: Border.all(
        color: Colors.grey,
        width: 1.5,
      ),
    );
  }

  Decoration circular() {
    return BoxDecoration(
      shape: BoxShape.circle,
      color: color,
      border: Border.all(
        color: Colors.grey,
        width: 1.5,
      ),
    );
  }

  Decoration rounded() {
    return BoxDecoration(
      borderRadius: BorderRadius.all(Radius.circular((size / 100) * 25)),
      color: color,
      border: Border.all(
        color: Colors.grey,
        width: 1.5,
      ),
    );
  }
}
